### PR TITLE
Added dashcam mode to video_recorder

### PR DESCRIPTION
**Pull request by IvanVN**
Added an optional "dashcam mode" for video recording. It will store a sequence of videos in a loop, deleting the oldest one.
Several parameters have been added:
- dashcam_mode: (bool) Enabled/disabled dashcam mode.
- video_length: (double) Length in minutes of each video in dashcam mode.
- n_videos: (int) Number of videos in the loop of the dashcam mode.
- video_topic: (string) input image topic
- use_posix_timestamp: (bool) Use posix formated (human readable) timestamps in filenames.

### CODE REVIEW

**Comment by JWhitleyWork** [comment]
I know this is old (just over a year) but I like the idea. However, I would rather call it something like "rolling_buffer," "ring_buffer," or similar instead of "dashcam" which is much more specific and, IMO, has other implications. Opinions welcome from the community (and maintainers - @SteveMacenski / @mjcarroll )

@IvanVN - If you're still keen on this type of thing, would you mind rebasing on `melodic`?

**Comment by mjcarroll** [comment]
Agree, I think that this is an interesting and useful feature.

I agree that the name should also be changed.  "rolling_buffer" seems like a suitable name.  I believe that the GoPro calls it looping.

One feature that would be nice to have is to have a service/topic to trigger saving the current contents and starting a new loop.  That way you could use something else in the system (say a failure incident or something) to cause it to store the duration of the video up to when the incident occurs.

**Comment by SteveMacenski** [comment]
I agree with @mjcarroll. Rolling buffer would make the most sense to me, and the ability to save sections of it would be useful. However I wouldnt make that feature as a requirement for merging this PR, but rather a feature request on a new ticket after this has been rebased and merged.

**Comment by IvanVN** [comment]
@JWhitleyAStuff I'll try to do it. Sadly right know I've to do it in my spare time and I don't have any melodic setup to test it, so I guess it will take me some time. If anyone wants to do it in the meantime, please feel free to do so.

**Comment by SteveMacenski** [comment]
@IvanVN Kinetic and Melodic are essentially the same for all intents and purposes for your PR. If you make sure it compiles and works in Kinetic and it passes Melodic CI, I think its fine.

**Comment by IvanVN** [comment]
I think it should be done. Compiled and tested in Kinetic. It also compiled OK in a Melodic docker.

File: image_view/src/nodes/video_recorder.cpp
Comment (JWhitleyWork) [inline image_view/src/nodes/video_recorder.cpp]: Can you please prefix the parameters that are only used for rolling_buffer_mode with `rb_` or `rbf_` to clarify that they aren't generic (e.g. `video_length` could be interpreted as "length of video to record" for any use of this node).

**Review by JWhitleyWork — CHANGES_REQUESTED** [review]
See comment about parameters. I've looked over it more thoroughly and this should be the last change request from me before approval. Thanks!

File: image_view/src/nodes/video_recorder.cpp
Comment (IvanVN) [reply image_view/src/nodes/video_recorder.cpp]: Done at 690da7f :+1:

File: image_view/src/nodes/video_recorder.cpp
Comment (SteveMacenski) [inline image_view/src/nodes/video_recorder.cpp]: Can you fix  this typo?

File: image_view/src/nodes/video_recorder.cpp
Comment (SteveMacenski) [inline image_view/src/nodes/video_recorder.cpp]: I'm a super huge not-fan of this. This is a global variable of variable size. The other variables while changing probably cannot cause additional memory allocation, but this will constantly. This shouldn't be global or should be set at runtime to a constant unchanging size.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: image_view/src/nodes/video_recorder.cpp
Comment (IvanVN) [reply image_view/src/nodes/video_recorder.cpp]: Sorry, but my c++ skills are pretty basic and I'm not sure of how to exactly do what you suggest. I would need some help on this.

I guess that it would be enough to just do `filenames.reserve(rb_n_videos + 1)` once we read the `rb_n_videos` parameter for the `std:vector`. However, there is no way (that I know of) of reserving also the size of the stored `std::string`, right?

Second option that comes to my mind is to use something like a non global `std::array<char[filename.size()], rb_n_videos>  filenames` and use it as input/output parameter to `generateStampedFilename` function. But I really need to check if this can be done (and how).

Anyway (and probably I'll be beaten up for this comment), in the worst case you would have `rb_n_videos + 1` memory allocations... Is it really so much of a big deal? :sweat_smile: I though this was exactly the purpose of the `vector` type.

Comment (JWhitleyWork) [reply image_view/src/nodes/video_recorder.cpp]: @IvanVN - Just a couple of suggestions/question here:

1. How are you actually using this filenames list? It looks like you use a function called `remove()` on the first filename in the vector but I don't exactly know what this function is doing or where it is included.
2. This should probably be a `queue` instead of a `vector`. A vector is not designed for efficiently removing items from the front of the list. A `queue` is exactly for this - it's an implementation of a FIFO (first-in-first-out) buffer.
3. If you reserve the amount of memory ahead of time equal to `rb_n_videos`, then that's the majority of the allocation done up-front. No, it won't reserve the size of each `std::string` but a vector is essentially a list of pointers to the actual storage locations for the elements anyway so the `reserve(N)` function will reserve the memory for one pointer for each element N. The pointed-to elements may change in size or location but the allocated memory for each pointer remains the same unless you add N+1 elements.
4. Rather than allocating `rb_n_videos` + 1, you could check the size of the container before adding the new element.

Comment (SteveMacenski) [reply image_view/src/nodes/video_recorder.cpp]: All above work. I know this pipeline may be used on some of the delivery robots and those guys are more concerned about that type of thing than your typical indoor robot (but far less than the autonomous car people). Just making sure we're not introducing a changing memory allocation where there wasn't one before would be good. The other global variables are statics so on runtime they allocate once and just hang out, this buffer changes size which is what I'm wanting to resolve.

**Comment by SteveMacenski** [comment]
What's the status of this @IvanVN

**Comment by IvanVN** [comment]
> What's the status of this @IvanVN

Sorry, I haven't have the time yet to look into this...

**Comment by JWhitleyWork** [comment]
@IvanVN Could I please get an update on this (and a rebase) if you're still interested in merging it? Sorry it's been forever since I checked back on it.

**Comment by IvanVN** [comment]
> @IvanVN Could I please get an update on this (and a rebase) if you're still interested in merging it? Sorry it's been forever since I checked back on it.

I've replaced the vector for a queue, as suggested. I'm having troubles rebasing to noetic, though. It doesn't compile because of some missing package (my current setup is melodic). I will come back to this in a couple of days and test it in a noetic docker container.

**Comment by JWhitleyWork** [comment]
@IvanVN Given that this has been hanging out for 2+ years, I'm going to close it as incomplete. If you come back to this at some time in the future, please re-open or create a new PR.

**Pull request closed without merging**
